internal/expr/prepare.go

```diff
@@ -0,0 +1,86 @@
+package expr
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+// PreparedExpr contains an SQL expression that is ready for execution.
+type PreparedExpr struct {
+	inputs []*inputPart
+	SQL    string
+}
+
+type typeNameToInfo map[string]*info
+
+func getKeys(m map[string]*info) []string {
+	i := 0
+	keys := make([]string, len(m))
+	for k := range m {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+// prepareInput checks that the input expression corresponds to a known type.
+func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	info, ok := ti[p.source.prefix]
+	if !ok {
+		return fmt.Errorf(`type %s unknown, have: %s`, p.source.prefix, strings.Join(getKeys(ti), ", "))
+	}
+
+	if _, ok = info.tagToField[p.source.name]; !ok {
+		return fmt.Errorf(`type %s has no %q db tag`,
+			info.structType.Name(), p.source.name)
+	}
+
+	return nil
+}
+
+// Prepare takes a parsed expression and all the Go objects mentioned in it.
+// The IO parts of the statement are checked for validity against the types
+// and expanded if necessary.
+func (pe *ParsedExpr) Prepare(args ...any) (expr *PreparedExpr, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot prepare expression: %s", err)
+		}
+	}()
+
+	var ti = make(typeNameToInfo)
+
+	// Generate and save reflection info.
+	for _, arg := range args {
+		info, err := typeInfo(arg)
+		if err != nil {
+			return nil, err
+		}
+		ti[info.structType.Name()] = info
+	}
+
+	var sql bytes.Buffer
+	// Check and expand each query part.
+
+	ins := []*inputPart{}
+
+	for _, part := range pe.queryParts {
+		switch p := part.(type) {
+		case *inputPart:
+			err := prepareInput(ti, p)
+			if err != nil {
+				return nil, err
+			}
+			sql.WriteString("?")
+			ins = append(ins, p)
+		case *outputPart:
+		case *bypassPart:
+			sql.WriteString(p.chunk)
+		default:
+			return nil, fmt.Errorf("internal error")
```

Comment:
This should say what the internal error is: ("internal error: unknown query part type: %T", part)

Reply:
Changed.